### PR TITLE
Add dockerfile packaging type

### PR DESCRIPTION
**Pull request by xiaoyu-work**
## Describe your changes

Add dockerfile packaging type. This packaging type will generate a Dockerfile that includes onnxruntime package and output models.

## Checklist before requesting a review
- [x] Add unit tests for this change.
- [x] Make sure all tests can pass.
- [x] Update documents if necessary.
- [x] Lint and apply fixes to your code by running `lintrunner -a`
- [ ] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.
- [ ] Is this PR including examples changes? If yes, please remember to update [example documentation](https://github.com/microsoft/Olive/blob/main/docs/source/examples.md) in a follow-up PR.

### CODE REVIEW

File: docs/source/features/packaging_output_models.md [hunk 1]
```diff
@@ -108,83 +117,92 @@ and for CPU, the best execution provider is CPUExecutionProvider, so the first r
...
-            Size of the mini-batch passed to each batch invocation. `10` by default.
-          * `extra_config [dict]`
-            Extra configurations for deployment. `None` by default.
-    * `include_sample_code [bool]`:
-      Whether or not to include sample code in zip file. Defaults to True
-    * `include_runtime_packages [bool]`:
-      Whether or not to include runtime packages (like onnxruntime) in zip file. Defaults to True
+  * `type [PackagingType]`:
+    Olive packaging type. Olive will package different artifacts based on `type`.
+  * `name [str]`:
```
Comment (devang-ml) [inline docs/source/features/packaging_output_models.md:136]: What is the default name for DockerFile type?

File: olive/engine/packaging/packaging_config.py [hunk 2]
```diff
@@ -36,6 +37,11 @@ class AzureMLModelsPackagingConfig(CommonPackagingConfig):
     description: Optional[str] = None
 
 
+class DockerfilePackagingConfig(CommonPackagingConfig):
+    base_image: str = "mcr.microsoft.com/azureml/openmpi4.1.0-cuda11.6-cudnn8-ubuntu20.04"
```
Comment (devang-ml) [inline olive/engine/packaging/packaging_config.py:41]: Please make this string a constant in olive/constants.py

File: docs/source/features/packaging_output_models.md (hunk 1 above)
Comment (xiaoyu-work) [reply docs/source/features/packaging_output_models.md:136]: There is no name for Dockerfile type. I can add this explanation to the doc.

File: olive/engine/packaging/packaging_config.py (hunk 2 above)
Comment (xiaoyu-work) [reply olive/engine/packaging/packaging_config.py:41]: Will do.

File: olive/common/constants.py [hunk 3]
```diff
@@ -4,9 +4,17 @@
...
+
+
+############# Engine #############
+
+DEFAULT_WORKFLOW_ID = "default_workflow"
+
+
+############# Packaging #############
+
+BASE_IMAGE = "mcr.microsoft.com/azureml/openmpi4.1.0-cuda11.6-cudnn8-ubuntu20.04"
```
Comment (jambayk) [inline olive/common/constants.py:20]: The latest ort packages don't support cuda 11.6 anymore. Might have to swich to `"mcr.microsoft.com/azureml/openmpi4.1.0-cuda11.8-cudnn8-ubuntu22.04"` like in https://github.com/microsoft/Olive/pull/1232

Comment (xiaoyu-work) [reply olive/common/constants.py]: Updated

**Review by jambayk — APPROVED** [review]

File: olive/engine/packaging/packaging_generator.py [hunk 4]
```diff
@@ -52,12 +53,65 @@
...
+    best_node = _get_best_candidate_node(pf_footprints, footprints)
+
+    docker_context_path = "docker_content"
+    content_path = output_dir / docker_context_path
+    if content_path.exists():
+        shutil.rmtree(content_path)
+    content_path.mkdir(parents=True)
+
+    if packaging_config.include_sample_code:
+        _package_sample_code(Path(__file__).parent, content_path, False)
```
[the remark this answers is not in the document]
Comment (jambayk) [reply olive/engine/packaging/packaging_generator.py]: does not apply

[the remark this answers is not in the document]
Comment (jambayk) [reply olive/engine/packaging/packaging_generator.py]: does not apply

**Pull request merged**